03cpp2/sec03Exceptions.md

```diff
@@ -0,0 +1,254 @@
+
+double g(double x)
+{
+    if(x <= 0)
+    {
+        throw std::logic_error("g(x) not defined for x <= 0");
+    }
+    return log(x);
+}
+
+double h(double x)
+{
+    if(x < 0)
+    {
+        throw std::logic_error("g(x) not defined for x < 0");
+    }
+    return sqrt(x);
+}
+
+double p(double x)
+{
+    return x*x - 3*x + 12;
+}
+
+double f(double x)
+{
+    return g(p(x)) * h((p(x)));
+}
+
+int main()
+{
+    bool complete = false;
+    double x = 0;
+
+    cout << "Enter a number." << endl;
+    cin >> x;
+
+    while(!complete) 
+    {
+      try 
+      {
+        double y = f(x);
+        cout << y << endl;
+        complete = true;
+      } 
+      catch(std::exception &e) 
+      {
+        cout << e.what() << endl;
+        cout << "Please enter a different number." << endl;
+        cin >> x;
+      }
+    }
+
+    return 0;
+}
+```
+
+## Throwing Exceptions
+
+We can also throw exceptions from our own functions, which allows code which calls our functions to handle errors that might occur within our function code. For example, we might have a function which calculates the inner product of two vectors. In this case, the length of both vectors must be the same for this to make sense! So we could write code like this:
+```cpp
+double InnerProduct(const vector<double> &x, const vector<double> &y)
+{
+    if(x.size() != y.size())
+    {
+        std::string errorMessage = "Inner product vectors different sizes: " + std::to_string(x.size()) + " and " + std::to_string(y.size()); 
+        throw std::range_error(errorMessage);
+    }
+    
+    double product = 0;
+    for(size_t i = 0; i < x.size(); i++)
+    {
+        product += x[i] * y[i];
+    }
+
+    return product;
+}
+```
+This will throw a `range_error` if the two vectors are not the same size. The code which calls this function can catch this exception using `catch(std::range_error e){}`. 
+```cpp
+int main()
+{
+    vector<double> a = {0.2, 0.1, 1.2, 5.99};
+    vector<double> b = {0.1, 1.8, 2.9};
+
+    double ab;
+    try
+    {
+        ab = InnerProduct(a, b);
+    }
+    catch(const std::range_error& e)
+    {
+        std::cerr << e.what() << '\n';
+    }
+    
+
+    return 0;
+}
+```
+
+**Warning** You can actually throw _any_ type in C++, it doesn't have to inherit from `exception`. The compiler will not complain if you throw, for example, an `int` or a `string`. To do so is bad practice, as these objects are not designed to be used to carry error information and codes which may call your functions will not be expecting them. This means calling code is unlikely to check for them, which will allow them to pass to the top level uncaught and halt execution. Restricting yourself to throwing classes which inherit from `exception` will make your code easier to understand for others, and compatible with other code. If you need values of other types to be reported with your exception then you can include them as member variables in you own exception class (see below). 
```

Comment:
**Warning**, **N.B.** and possibly other tips should be checked for consistency. It's a great way to signal things to look out for, we just need to make sure a student understands the hierarchy of tips.